openedx_tagging/core/tagging/api.py

```diff
@@ -107,3 +106,68 @@ def tag_object(
     """
 
     return taxonomy.tag_object(tags, object_id, object_type)
+
+
+def autocomplete_tags(taxonomy: Taxonomy, prefix: str, object_id: str= None, count=10) -> QuerySet:
+    """
+    Returns the first `count` tag values in the given Taxonomy with names 
+    that begin with the given prefix string. The result is returned in alphabetical
+
+    Closed taxonomies return tag values that exist in the taxonomy. Also excludes all Tags that the
+    `object_id` already has. 
+
+    Free-text taxonomies return only tag values that are currently in use on (another) object. Also excludes
+    the tags used by `object_id`.
+    """
+    result = []
+
+    # Fetch tags that the object already has to exclude them from the result
+    excluded_tags = ObjectTag.objects.filter(object_id=object_id, taxonomy=taxonomy)
+    if taxonomy.allow_free_text:
+        # Free-text taxonomy
+
+        # Obtain the value of the excluded tags
+        excluded_tags = excluded_tags.values_list('_value', flat=True)
+        return (
+            # Fetch object tags from this taxonomy whose value starts with the given prefix
+            ObjectTag.objects.filter(taxonomy=taxonomy, _value__istartswith=prefix)
+            # omit any free-text tags whose values match the tags on the given object
+            .exclude(_value__in=excluded_tags)
+            # alphabetical ordering
+            .order_by('_value')
+            # obtain the values of the tags
+            .values_list('_value', flat=True)
+            # remove repeats
+            .distinct()
+            # get only first `count` values
+            [:count]
+        )
+    else:
+        # Closed taxonomy
+        
+        # Obtain the id of the excluded tags
+        excluded_tags = excluded_tags.filter(tag__isnull=False).values_list('tag__id', flat=True)
+        return (
+            # Fetch tags from this taxonomy whose value starts with the given prefix
+            Tag.objects.filter(taxonomy=taxonomy, value__istartswith=prefix)
+            # omit any tags applied to the given object
+            .exclude(id__in=excluded_tags)
+            # alphabetical ordering
+            .order_by('value')
+            # obtain only id and values of the tags
+            .values('id', 'value')
+            # get only first `count` values
+            [:count]
+        )
+
+
+def autocomplete_tags_result(taxonomy: Taxonomy, prefix: str, object_id: str= None, count=10) -> List[TagResult]:
+    """
+    Calls `autocomplete_tags` and serialize the results into `TagResult`
+    """
+    result = autocomplete_tags(taxonomy, prefix, object_id=object_id, count=count)
+    if taxonomy.allow_free_text:
+        result = [TagResult(id=None, value=value) for value in result]
+    else:
+        result = [TagResult(id=item['id'], value=item['value']) for item in result]
+    return result
```

Comment:
@ChrisChV I think we can let the REST API serialize the results here, so this method doesn't need to be in the python API.

Reply:
Updated here https://github.com/openedx/openedx-learning/pull/59/commits/223b3a4c314d84255a36e73cd80d9b108e577568